kernel/arch/dreamcast/hardware/maple/controller.c

```diff
@@ -11,22 +11,48 @@
 #include <string.h>
 #include <assert.h>
 
+/* Location of controller capabilities within function_data array */
+#define CONT_FUNCTION_DATA_INDEX  0 
+
+/* Raw controller condition structure */
+typedef struct cont_cond {
+    uint16_t buttons;  /* buttons bitfield */
+    uint8_t rtrig;     /* right trigger */
+    uint8_t ltrig;     /* left trigger */
+    uint8_t joyx;      /* joystick X */
+    uint8_t joyy;      /* joystick Y */
+    uint8_t joy2x;     /* second joystick X */
+    uint8_t joy2y;     /* second joystick Y */
+} cont_cond_t;
+
 static cont_btn_callback_t btn_callback = NULL;
-static uint8 btn_callback_addr = 0;
-static uint32 btn_callback_btns = 0;
+static uint8_t btn_callback_addr = 0;
+static uint32_t btn_callback_btns = 0;
+
+/* Check whether the controller has EXACTLY the given capabilties. */
+int cont_is_type(const maple_device_t *cont, uint32_t type) {
+    return cont->info.function_data[CONT_FUNCTION_DATA_INDEX] == type; 
```

Comment:
Should these check for derefs, and -1 on bad input vs 0/1 for results?

Reply:
Added!